tensorflow/python/estimator/canned/boosted_trees.py

```diff
@@ -910,8 +939,105 @@ def _create_regression_head(label_dimension, weight_column=None):
+
+  Raises:
+    AssertionError: If feature importances contain negative value.
+      Or if normalize = True and normalization is not possible
+      (e.g. ensemble is empty or trees contain only a root node).
+  """
+  tree_importances = [_compute_feature_importances_per_tree(tree, num_features)
+                      for tree in tree_ensemble.trees]
+  tree_importances = np.array(tree_importances)
+  tree_weights = np.array(tree_ensemble.tree_weights).reshape(-1, 1)
+  feature_importances = np.sum(tree_importances * tree_weights, axis=0)
+  assert np.all(feature_importances >= 0), ('feature_importances '
+                                            'must be non-negative.')
+  if normalize:
+    normalizer = np.sum(feature_importances)
+    assert normalizer > 0, 'Trees are all empty or contain only a root node.'
+    feature_importances /= normalizer
+
+  sorted_feature_idx = np.argsort(feature_importances)[::-1]
+  return sorted_feature_idx, feature_importances[sorted_feature_idx]
+
+
+class _BoostedTrees(estimator.Estimator):
+  """Base class for boosted trees estimators."""
+
+  def __init__(self, model_fn, model_dir, config, feature_columns):
+    super(_BoostedTrees, self).__init__(
+        model_fn=model_fn, model_dir=model_dir, config=config)
+
+    self._sorted_feature_columns = sorted(feature_columns,
+                                          key=lambda tc: tc.name)
+    self._num_features = _calculate_num_features(self._sorted_feature_columns)
+    self._names_for_feature_id = np.array(
+        _generate_feature_name_mapping(self._sorted_feature_columns))
+
+  def experimental_feature_importances(self, normalize=False):
+    """Computes gain-based feature importances.
+
+    The higher the value, the more important the corresponding feature.
+
+    Args:
+      normalize: If True, normalize the feature importances.
+
+    Returns:
+      sorted_feature_names: 1-D array of feature name which is sorted
+        by its feature importance.
+      feature_importances: 1-D array of the corresponding feature importance.
+
+    Raises:
+      ValueError: When attempting to normalize on an empty ensemble
+        or an ensemble of trees which have no splits.
+    """
+    reader = checkpoint_utils.load_checkpoint(self._model_dir)
+    serialized = reader.get_tensor('boosted_trees:0_serialized')
+    if not serialized:
+      raise ValueError('Found empty serialized string for TreeEnsemble.'
+                       'You should only call this method after training.')
+    ensemble_proto = boosted_trees_pb2.TreeEnsemble()
+    ensemble_proto.ParseFromString(serialized)
+
+    sorted_feature_id, importances = _compute_feature_importances(
+        ensemble_proto, self._num_features, normalize)
+    return self._names_for_feature_id[sorted_feature_id], importances
+
+
 @estimator_export('estimator.BoostedTreesClassifier')
-class BoostedTreesClassifier(estimator.Estimator):
+class BoostedTreesClassifier(_BoostedTrees):
```

Comment:
can we dissolve the inheritance here? You can convert the experimental_feature_importances to a helper function and call it from all other boosted tree estimators.

Reply:
I agree with @nataliaponomareva . By the way, because _BoostedTrees is a private class, I think we can remove it easily if we find a better solution in the future. How do you think?

Reply:
i'm okay with it.